src/drive/actions/async.js

```diff
@@ -148,101 +169,112 @@ class Stack {
+      updated_at: byUpdatedAt,
+      size: bySize,
+      recentFiles: recentFiles
+    }
   }
 
-  getFolder = async folderId => {
+  normalizeFileFromPouchDB = f => ({
+    ...f,
+    id: f._id,
+    _type: 'io.cozy.files'
+  })
+
+  getFolder = async (folderId, specialFolders = []) => {
     const db = cozy.client.offline.getDatabase('io.cozy.files')
     const folder = await db.get(folderId)
     const parent = !!folder.dir_id && (await db.get(folder.dir_id))
-    const files = await this.getFolderContents(folderId)
+    const recentsFirst =
+      !!parent &&
+      shouldShowRecentsFirst(folder.path, parent.path, specialFolders)
+
+    const files = recentsFirst
+      ? await this.getSortedFolder(folderId, 'updated_at', 'desc')
+      : await this.getSortedFolder(folderId, 'name', 'asc')
+
     return {
-      ...normalizeFileFromPouchDB(folder),
+      ...this.normalizeFileFromPouchDB(folder),
       contents: {
         data: files,
         meta: {}
       },
-      parent: !!parent && normalizeFileFromPouchDB(parent)
+      parent: !!parent && this.normalizeFileFromPouchDB(parent)
     }
   }
 
-  getFolderContents = async (folderId, skip = 0, limit = FILES_FETCH_LIMIT) => {
-    const db = cozy.client.offline.getDatabase('io.cozy.files')
-    const resp = await db.find({
-      selector: {
-        dir_id: folderId,
-        name: { $gte: null },
-        type: { $gte: null }
-      },
-      use_index: this.folderIndex,
-      sort: ['dir_id', { type: 'desc' }, { name: 'desc' }],
-      limit,
-      skip
-    })
-    const files = resp.docs
-      .filter(f => f._id !== TRASH_DIR_ID)
-      .map(f => normalizeFileFromPouchDB(f))
-    return files
-  }
+  getFolderContents = (folderId, skip = 0, limit = FILES_FETCH_LIMIT) =>
+    this.getSortedFolder(folderId, 'name', 'asc', skip, limit)
 
   getSortedFolder = async (
     folderId,
     sortAttribute,
     sortOrder = 'asc',
     skip = 0,
-    limit = FILES_FETCH_LIMIT,
-    loadedFoldersCount = 0,
-    loadedFilesCount = 0
+    limit = FILES_FETCH_LIMIT
   ) => {
-    const shouldSeparateFilesAndFolders = sortAttribute === 'name'
-    const index = this.sortIndexes[sortAttribute]
-    if (!index) {
-      throw new Error(`Can't sort on ${sortAttribute}`)
-    }
-    const db = cozy.client.offline.getDatabase('io.cozy.files')
-    const query = (selector = {}, skipRows = skip) =>
-      db
-        .find({
-          selector: {
-            ...selector,
-            ...{
-              dir_id: folderId,
-              _id: { $ne: TRASH_DIR_ID },
-              [sortAttribute]: { $gte: null }
-            }
-          },
-          use_index: index,
-          sort: [{ [sortAttribute]: sortOrder }],
-          skip: skipRows,
-          limit
-        })
-        .then(resp => resp.docs.map(f => normalizeFileFromPouchDB(f)))
-
-    if (shouldSeparateFilesAndFolders) {
-      const folders = await query({
-        type: 'directory'
+    const isFirstLoad = skip === 0
```

Comment:
You'll notice that the content of `getSortedFolder` is literally the same between the StackAdapter and the PouchAdapter — only the underlying `query` method changes.

I'd like to avoid that duplication, but not sure how. They are both definitely class methods, but having one adapter `extend` the other doesn't seem right  either, as literally everything else changes. And having both `extend` a base class is maybe a bit overkill?

Reply:
I'm not sure trying to avoid that duplication is worth it right now. Maybe we'll find a better way later if we intergate this work in cozy-client.